python/cudf/cudf/core/scalar.py

```diff
@@ -358,7 +358,7 @@ def _dispatch_scalar_unaop(self, op):
         return getattr(self.value, op)()
 
     def astype(self, dtype):
-        return Scalar(self.device_value, dtype)
+        return Scalar(self.value, dtype)
```

Comment:
This is fundamentally changing the behavior of this function now. If the device_value was the only thing set this will now unnecessarily force computing the host value. I think we need something like:
```
if self._is_device_value_current:
    return Scalar(self.device_value, dtype)
else:
    return Scalar(self.value, dtype)
```

Reply:
Actually, I'm not sure this works as expected, in looking at the `__init__` of `Scalar`, it looks like if we pass a `DeviceScalar` it won't actually use the `dtype` passed in at all.

Hmm...

Reply:
In this case, `Scalar(self.device_value, dtype)` will always return a `Scalar` with `dtype==None` (see ctor). If user passes in a device_side_only scalar to `interval_range`, we might see an exception there.

Reply:
hmm okay, lets leave this as is for now then because this is the only way to get correct results from typecasting.

Reply:
https://github.com/rapidsai/cudf/issues/7835 issue raised here to keep track.